src/Servers/Kestrel/Core/src/Internal/Http/MessageBody.cs

```diff
@@ -165,16 +166,82 @@ protected ValueTask<ReadResult> StartTimingReadAsync(ValueTask<ReadResult> readA
             return readAwaitable;
         }
 
-        protected void StopTimingRead(long bytesRead)
+        protected void CountBytesRead(long bytesInReadResult)
         {
-            _context.TimeoutControl.BytesRead(bytesRead - _alreadyTimedBytes);
-            _alreadyTimedBytes = 0;
+            var numFirstSeenBytes = bytesInReadResult - _alreadyTimedBytes;
```

Comment:
@halter73  Should this then update `_alreadyTimedBytes`?

```csharp
var numFirstSeenBytes = bytesInReadResult - _alreadyTimedBytes;

if (numFirstSeenBytes > 0)
{
    _context.TimeoutControl.BytesRead(numFirstSeenBytes);
    _alreadyTimedBytes = bytesInReadResult ;
}
```

Reply:
Ah, nvm I think its ok

Reply:
Yeah. We count all the bytes in the ReadResult towards rate calculations, examined or not. Then in Advance, if we don't consume everything, we count all the unconsumed *and* unexamined bytes towards _alreadyTimedBytes, since both the unconsumed and unexamined bytes will be present in the next read result. What got you looking at this?

Reply:
Was getting test failures with https://github.com/aspnet/AspNetCore/pull/11942#discussion_r315836996 when I removed all the changes that weren't directly related to the streams; as assumed this superseded them.

Issue wasn't with counting though; needed `TryStart`/`TryStop` in Http2MessageBody